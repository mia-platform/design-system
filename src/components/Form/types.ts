
export enum Layout {
  Horizontal = 'horizontal',
  Vertical = 'vertical'
}
