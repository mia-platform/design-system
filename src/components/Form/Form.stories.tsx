/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Fragment, ReactElement } from 'react'
import { Meta, type StoryObj } from '@storybook/react'
import { Flex } from 'antd'

import { BodyM } from '../Typography/BodyX/BodyM'
import { Button } from '../Button'
import { Card } from '../Card'
import { Checkbox } from '../Checkbox'
import { CheckboxGroup } from '../CheckboxGroup'
import { Form } from './Form.tsx'
import { Input } from '../Input'
import { InputNumber } from '../InputNumber'
import { RadioGroup } from '../RadioGroup'
import { Search } from '../Search'
import { Select } from '../Select'
import { Switch } from '../Switch'
import { TextArea } from '../TextArea'

const meta = {
  component: Form,
  args: {
    children: (
      <Fragment>
        <Form.Item name="firstName">
          <Input />
        </Form.Item>
        <Form.Item name="lastName">
          <Input />
        </Form.Item>
      </Fragment>
    ),
  },
  argTypes: {},
} satisfies Meta<typeof Form>

export default meta
type Story = StoryObj<typeof meta>

const options = [
  ...Array(3).keys(),
].map((id) => ({
  value: `value ${id + 1}`,
  label: `label ${id + 1}`,
}))

export const Default: Story = {
  args: {},
}

export const Horizontal: Story = {
  args: {
    layout: Form.Layout.Horizontal,
  },
}

export const WithInitialValues: Story = {
  args: {
    initialValues: {
      firstName: 'John',
      lastName: 'Doe',
    },
  },
}

export const WithCustomColumns = {
  args: {
    columns: 1,
  },
}

export const WithCustomGap: Story = {
  args: {
    columns: 1,
    gap: 160,
  },
}

export const WithCustomSubmit: Story = {
  args: {
    submitButton: ({ form }) => (
      <Flex gap={8}>
        <Button
          hierarchy={Button.Hierarchy.Danger}
          onClick={() => form?.resetFields()}
        >
          Clear fields
        </Button>
        <Form.SubmitButton />
      </Flex>
    ),
  },
}

export const WithDetachedSubmit = (): ReactElement => {
  const handleSubmit = (values: unknown): void => {
    alert(`onFinish:\n${JSON.stringify(values)}`)
  }
  return (
    <Flex align="center" justify="space-between">
      <Card>
        <Form id="my-form" submitButton={false} onFinish={handleSubmit}>
          <Form.Item name="firstName">
            <Input />
          </Form.Item>
          <Form.Item name="lastName">
            <Input />
          </Form.Item>
        </Form>
      </Card>
      <Card>
        <Form.SubmitButton form="my-form" />
      </Card>
    </Flex>
  )
}

export const WithValidation: Story = {
  args: {
    onFinishFailed: (reason: unknown) => {
      alert(`OnFinishFailed:\n${JSON.stringify(reason)}`)
    },
    children: (
      <Form.Item
        name="firstName"
        rules={[{ validator: () => Promise.reject(new Error('Validation error')) }]}
      >
        {() => <Input />}
      </Form.Item>
    ),
  },
}

export const ComplexForm: Story = {
  args: {
    columns: 3,
    initialValues: {
      input: 'input',
      textarea: 'textarea',
      number: 1,
      search: options[0].value,
      select: options[0].value,
      checkboxGroup: [options[0].value],
      inputAddon: { before: options[0].value, value: 'text' },
      switch: true,
      checkbox: true,
      radioGroup: options[0].value,
      custom: 0,
    },
    children: (
      <Fragment>
        <Form.Item isFullWidth name="inputAddon" rules={[{ required: true }]}>
          <Input addonBefore={{ type: Input.AddonType.Select, options }} />
        </Form.Item>
        <Form.Item name="textfield" rules={[{ required: true }]}>
          <Input />
        </Form.Item>
        <Form.Item name="textarea" rules={[{ required: true }]}>
          <TextArea />
        </Form.Item>
        <Form.Item name="number" rules={[{ required: true }]} >
          <InputNumber />
        </Form.Item>
        <Form.Item name="search" rules={[{ required: true }]} span={2}>
          <Search options={options} />
        </Form.Item>
        <Form.Item name="select" rules={[{ required: true }]} >
          <Select options={options} />
        </Form.Item>
        <Flex
          gap={16}
          justify="space-between"
          style={{ gridColumn: '-1 / 1' }}
        >
          <Form.Item name="checkbox" rules={[{ required: true }]} >
            <Checkbox />
          </Form.Item>
          <Form.Item name="switch" rules={[{ required: true }]} >
            <Switch />
          </Form.Item>
          <Form.Item name="checkboxGroup" rules={[{ required: true }]}>
            <CheckboxGroup
              direction={CheckboxGroup.Direction.Horizontal}
              options={options}
            />
          </Form.Item>
          <Form.Item name="radioGroup" rules={[{ required: true }]}>
            <RadioGroup
              direction={RadioGroup.Direction.Horizontal}
              options={options}
            />
          </Form.Item>
        </Flex>
        <Form.Item name="custom">
          {({ value, onChange }) => {
            const handleClick = (): void => {
              if (onChange) {
                onChange(Number(value) + 1)
              }
            }
            return (
              <Button onClick={handleClick}>{`clicked ${value} times`}</Button>
            )
          }}
        </Form.Item>
        <Form.Item
          name="password"
          rules={[Form.Validators.required('password is required')]}
        >
          <Input type={Input.Type.Password} />
        </Form.Item>
        <Form.Item
          name="confirmPassword"
          rules={[
            Form.Validators.required('confirm password is required'),
            Form.Validators.checkEquals('password', "password and confirm password don't match!"),
          ]}
        >
          <Input type={Input.Type.Password} />
        </Form.Item>
        <Form.Item isFullWidth label={<BodyM isBold>Form values:</BodyM>} shouldUpdate>
          {
            ({ form }) => (
              <Card>
                <pre>{JSON.stringify(form?.getFieldsValue(true), null, '\t')}</pre>
              </Card>
            )
          }
        </Form.Item>
      </Fragment>
    ),
  },
}
