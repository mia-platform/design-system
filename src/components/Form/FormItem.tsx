import { Form as AntForm, FormItemProps as AntFormItemProps } from 'antd'
import { ReactElement } from 'react'

export const FormItem = ({ name, ...props }: AntFormItemProps): ReactElement => {
  return (
    <AntForm.Item label={name} name={name} {...props} />
  )
}
