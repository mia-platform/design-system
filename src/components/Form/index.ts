
export { FormItem } from './FormItem/FormItem.tsx'
export { Form } from './Form.tsx'

