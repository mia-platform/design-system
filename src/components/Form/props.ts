/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { FormItemProps as AntFormItemProps, FormProps as AntFormProps, FormInstance } from 'antd'
import { CSSProperties, ReactElement, ReactNode } from 'react'

import { Layout } from './types.ts'

export type RenderProps = {
  form?: FormInstance,
  value?: unknown;
  onChange?: (value: unknown) => void
}

export type FormProps<Values extends Record<string, unknown>> = {

  /**
   * A unique identifier for the form.
   */
  id?: string;

  /**
   * The content of the form, typically form fields or other React elements.
   */
  children?: ReactNode;

  /**
   * The number of columns for form layout. Determines how form fields are arranged.
   */
  columns?: number;

  /**
   * The gap (in pixels) between form fields.
   */
  gap?: number;

  /**
   * Defines the layout direction of the form. Can be 'horizontal' or 'vertical'.
   */
  layout?: Layout

  /**
   * Custom styles for the form, passed as a CSSProperties object.
   */
  style?: CSSProperties;

  /**
   * The initial values for the form fields.
   */
  initialValues?: Values;

  /**
   * Specifies whether to retain field values when the field is removed from the form. Defaults to `true`.
   */
  preserve?: boolean;

  /**
   * Renders a submit button. Can be a boolean (to toggle default button), a string (button text),
   * or a function that receives the form instance and returns a custom ReactNode.
   */
  submitButton?: boolean | string | ((props: RenderProps) => ReactNode);

  /**
   * Callback triggered when the form values change.
   */
  onValuesChange?: AntFormProps<Values>['onValuesChange'];

  /**
   * Callback triggered when the form is successfully submitted.
   */
  onFinish?: AntFormProps<Values>['onFinish'];

  /**
   * Callback triggered when form submission fails validation.
   */
  onFinishFailed?: AntFormProps<Values>['onFinishFailed'];
}

export type FormItemProps = {

  /**
   * The name of the form field, used for form data management.
   */
  name?: string;

  /**
   * The label for the field, displayed next to or above the form element.
   */
  label?: ReactNode;

  /**
   * Inline style for the FormItem container.
   */
  style?: CSSProperties;

  /**
   * The number of grid columns the FormItem should span.
   */
  span?: number;

  /**
   * Alignment of the content within the FormItem.
   */
  justify?: 'start' | 'center' | 'end';

  /**
   * Whether the FormItem should take the full width of its container.
   */
  isFullWidth?: boolean;

  /**
   * Validation rules for the form field.
   */
  rules?: AntFormItemProps['rules'];

  /**
   * The property name used for the value of the form element (e.g., `checked` for a Checkbox).
   */
  valuePropName?: AntFormItemProps['valuePropName'];

  /**
   * Function to extract the value from the event generated by the form element.
   */
  getValueFromEvent?: AntFormItemProps['getValueFromEvent'];

  /**
   * Content of the FormItem; can be a React element or a render function.
   */
  children: ReactElement | ((props: RenderProps) => ReactNode);

  /**
   * Whether the FormItem should update when the state of other fields changes.
   */
  shouldUpdate?: boolean;

  /**
   * List of fields that this FormItem depends on, used to trigger conditional updates.
   */
  dependencies?: string[];

  /**
   * Help text displayed below the form field.
   */
  help?: ReactNode;

  /**
   * Additional content displayed below the FormItem, similar to `help` but for more general purposes.
   */
  extra?: ReactNode;

  /**
   * Whether the field is required, showing an asterisk and applying validation.
   */
  required?: boolean;
};

