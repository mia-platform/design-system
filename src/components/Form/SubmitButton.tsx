import { ReactElement, ReactNode, useCallback } from 'react'
import { FormInstance } from 'antd'

import { Button } from '../Button'
import { FormItem } from './FormItem.tsx'

export type SubmitButtonProps = {
  children?: ReactNode | ((form: FormInstance) => ReactNode)
}

const defaultSubmitButton = (label: ReactNode = 'Submit'): ReactNode => {
  return (
    <Button htmlType={Button.HTMLType.Submit}>
      {label}
    </Button>
  )
}

export const SubmitButton = (
  {
    children,
  }: SubmitButtonProps
): ReactElement => {
  const renderSubmitButton = useCallback(({ form }: { form: FormInstance }) => {
    if (typeof children === 'function') {
      return children(form)
    }
    if (typeof children === 'boolean') {
      return children && defaultSubmitButton()
    }
    return defaultSubmitButton(children)
  }, [children])

  return (
    <FormItem isFullWidth justify="end">
      {renderSubmitButton}
    </FormItem>
  )
}
