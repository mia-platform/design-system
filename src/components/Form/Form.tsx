import { ReactElement, useMemo } from 'react'
import { Form as AntForm } from 'antd'

import * as Validators from './Validators'
import { FormProps } from './props.ts'
import { Layout } from './types.ts'
import { SubmitButton } from './SubitButton/SubmitButton.tsx'

const defaults = {
  columns: 2,
  gap: 24,
  layout: Layout.Vertical,
  submitButton: true,
}

export const Form = <Values extends Record<string, unknown>, >({
  children,
  style: styleProp,
  submitButton: submitButtonProp = defaults.submitButton,
  layout = defaults.layout,
  columns = defaults.columns,
  gap = defaults.gap,
  id,
  initialValues,
  preserve,
  onValuesChange,
  onFinish,
  onFinishFailed,
}: FormProps<Values>): ReactElement => {
  const style = useMemo(() => ({
    display: 'grid',
    gridTemplate: `auto / repeat(${columns}, 1fr)`,
    gap: `${gap}px`,
    ...styleProp,
  }), [columns, gap, styleProp])

  return (
    <AntForm<Values>
      id={id}
      initialValues={initialValues}
      layout={layout}
      preserve={preserve}
      style={style}
      onFinish={onFinish}
      onFinishFailed={onFinishFailed}
      onValuesChange={onValuesChange}
    >
      {children}
      {
        submitButtonProp && (
          <SubmitButton>
            {submitButtonProp}
          </SubmitButton>
        )
      }
    </AntForm>
  )
}

Form.Layout = Layout
Form.SubmitButton = SubmitButton
Form.Validators = Validators
