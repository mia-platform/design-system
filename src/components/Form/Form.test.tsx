/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Fragment, ReactElement } from 'react'
import { screen, waitFor } from '@testing-library/react'
import { fireEvent } from '@testing-library/dom'

import { Button } from '../Button/Button.tsx'
import { Form } from './Form.tsx'
import { Input } from '../Input'
import { render } from '../../test-utils.tsx'

const exampleText = 'text'

const props = {
  children: (
    <Fragment>
      <Form.Item name="firstName" rules={[Form.Validators.required()]}>
        <Input />
      </Form.Item>
      <Form.Item name="lastName" rules={[Form.Validators.required()]}>
        <Input />
      </Form.Item>
    </Fragment>
  ),
}

describe('Form Component', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  test('renders correctly', async() => {
    const { asFragment } = render(<Form {...props} />)
    await waitFor(() => expect(asFragment()).toMatchSnapshot())
  })

  test('renders horizontal correctly', async() => {
    const { asFragment } = render(<Form layout={Form.Layout.Horizontal} {...props} />)
    await waitFor(() => expect(asFragment()).toMatchSnapshot())
  })

  test('renders custom gap correctly', async() => {
    const { asFragment } = render(<Form gap={64} {...props} />)
    await waitFor(() => expect(asFragment()).toMatchSnapshot())
  })

  test('renders custom columns count correctly', async() => {
    const { asFragment } = render(<Form columns={3} {...props} />)
    await waitFor(() => expect(asFragment()).toMatchSnapshot())
  })

  test('renders initialValues correctly', async() => {
    const { asFragment } = render(<Form initialValues={{ firstName: 'foo', lastName: 'bar' }} {...props} />)
    await waitFor(() => expect(asFragment()).toMatchSnapshot())
  })

  test('renders with custom submit button label', async() => {
    const { asFragment } = render(<Form submitButton={exampleText} {...props} />
    )
    await waitFor(() => expect(asFragment()).toMatchSnapshot())
    expect(screen.getByText(exampleText)).toBeInTheDocument()
  })

  test('renders with custom submit button function', async() => {
    const submitButton = () : ReactElement => {
      return <Form.SubmitButton>{exampleText}</Form.SubmitButton>
    }
    const { asFragment } = render(<Form submitButton={submitButton} {...props} />
    )
    await waitFor(() => expect(asFragment()).toMatchSnapshot())
    expect(screen.getByText(exampleText)).toBeInTheDocument()
  })

  test('renders with custom submit button function', async() => {
    const submitButton = () : ReactElement => {
      return <Form.SubmitButton>{exampleText}</Form.SubmitButton>
    }
    const { asFragment } = render(<Form submitButton={submitButton} {...props} />
    )
    await waitFor(() => expect(asFragment()).toMatchSnapshot())
    expect(screen.getByText(exampleText)).toBeInTheDocument()
  })

  test('should call onFinishFailed if the form is not valid', async() => {
    const onFinishFailed = jest.fn()
    render(<Form onFinishFailed={onFinishFailed} {...props} />
    )
    const button = screen.getByRole('button', { name: 'Submit' })
    fireEvent.click(button)

    await waitFor(async() => {
      expect(onFinishFailed).toHaveBeenCalled()
    })

    expect(await screen.findAllByText('The field is required')).toHaveLength(2)
  })

  test('should fails test using instance if the form is not valid', async() => {
    const InlineForm = (): ReactElement => {
      const [form] = Form.useForm()
      const id = 'the-form'
      const onClick = (): void => {
        form.validateFields()
          .then(() => {
            throw new Error('Promise should not resolve')
          })
          .catch(({ errorFields }) => {
            expect(errorFields).toEqual([
              { 'errors': ['The field is required'], 'name': ['firstName'], 'warnings': [] },
              { 'errors': ['The field is required'], 'name': ['lastName'], 'warnings': [] },
            ])
          })
      }

      return (
        <>
          <Form
            form={form}
            id={id}
            submitButton={false}
          >
            <Form.Item name="firstName" rules={[Form.Validators.required()]}>
              <Input />
            </Form.Item>
            <Form.Item name="lastName" rules={[Form.Validators.required()]}>
              <Input />
            </Form.Item>
          </Form>
          <Button onClick={onClick} >{'Submit'}</Button>
        </>
      )
    }

    render(<InlineForm />)
    const button = screen.getByRole('button', { name: 'Submit' })
    fireEvent.click(button)

    expect(await screen.findAllByText('The field is required')).toHaveLength(2)
  })

  test('should call onFinish if the form is valid', async() => {
    const onFinish = jest.fn()
    render(<Form onFinish={onFinish} {...props} />
    )
    const button = screen.getByRole('button', { name: 'Submit' })

    const firstNameInput = screen.getByRole('textbox', { name: /firstname/i })
    const lastNameInput = screen.getByRole('textbox', { name: /lastname/i })

    fireEvent.change(firstNameInput, { target: { value: exampleText } })
    fireEvent.change(lastNameInput, { target: { value: exampleText } })

    fireEvent.click(button)

    await waitFor(async() => {
      expect(onFinish).toHaveBeenCalledWith({ firstName: exampleText, lastName: exampleText })
    })
  })
})
