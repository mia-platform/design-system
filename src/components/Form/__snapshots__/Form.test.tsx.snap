// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormItem Component renders input FormItem correctly 1`] = `
<DocumentFragment>
  <form
    class="mia-platform-form mia-platform-form-vertical"
    style="display: grid; grid-template: auto / repeat(2, 1fr); gap: 24px;"
  >
    <div
      class="mia-platform-form-item"
      style="margin: 0px; justify-self: end; grid-column: -1 / 1;"
    >
      <div
        class="mia-platform-row mia-platform-form-item-row"
      >
        <div
          class="mia-platform-col mia-platform-form-item-control"
        >
          <div
            class="mia-platform-form-item-control-input"
          >
            <div
              class="mia-platform-form-item-control-input-content"
            >
              <button
                class="mia-platform-btn mia-platform-btn-primary mia-platform-btn-color-primary mia-platform-btn-variant-solid button buttonMd"
                type="submit"
              >
                <div
                  class="buttonText"
                >
                  Submit
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
